### PR TITLE
expression: implement vectorized evaluation for builtinAesEncryptSig

### PR DESCRIPTION
**Pull request by js00070**
PCP #12106 

### What problem does this PR solve? 
implement vectorized evaluation for builtinAesEncryptSig in #12106 

### What is changed and how it works?
about 10% faster
```
BenchmarkVectorizedBuiltinEncryptionFunc/builtinAesEncryptSig-VecBuiltinFunc-12             3082            374824 ns/op          397520 B/op       6955 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinAesEncryptSig-NonVecBuiltinFunc-12                  2662            410965 ns/op          407520 B/op       7470 allocs/op
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_encryption_vec.go [hunk 1]
```diff
@@ -366,11 +366,54 @@ func (b *builtinCompressSig) vecEvalString(input *chunk.Chunk, result *chunk.Col
...
+	defer b.bufAllocator.put(keyBuf)
+	if err := b.args[1].VecEvalString(b.ctx, input, keyBuf); err != nil {
+		return err
+	}
+
+	if b.modeName != "ecb" {
+		return errors.Errorf("unsupported block encryption mode - %v", b.modeName)
+	}
+
+	isWarning := !b.ivRequired && len(b.args) == 3
```
Comment (XuHuaiyu) [inline expression/builtin_encryption_vec.go]: Miss the original comment here.

File: expression/builtin_encryption_vec.go [hunk 2]
```diff
@@ -366,11 +366,54 @@ func (b *builtinCompressSig) vecEvalString(input *chunk.Chunk, result *chunk.Col
...
+	result.ReserveString(n)
+	for i := 0; i < n; i++ {
+		if strBuf.IsNull(i) || keyBuf.IsNull(i) {
+			result.AppendNull()
+			continue
+		}
+		if isWarning {
+			b.ctx.GetSessionVars().StmtCtx.AppendWarning(errWarnOptionIgnored.GenWithStackByArgs("IV"))
+		}
+		key := encrypt.DeriveKeyMySQL([]byte(keyBuf.GetString(i)), b.keySize)
```
Comment (XuHuaiyu) [inline expression/builtin_encryption_vec.go]: GetString ==> GetBytes

File: expression/builtin_encryption_vec.go [hunk 3]
```diff
@@ -366,11 +366,54 @@ func (b *builtinCompressSig) vecEvalString(input *chunk.Chunk, result *chunk.Col
...
+		cipherText, err := encrypt.AESEncryptWithECB([]byte(strBuf.GetString(i)), key)
```
Comment (XuHuaiyu) [inline expression/builtin_encryption_vec.go]: ditto

File: expression/builtin_encryption_vec.go [hunk 4]
```diff
@@ -366,11 +366,54 @@ func (b *builtinCompressSig) vecEvalString(input *chunk.Chunk, result *chunk.Col
...
+		if err != nil {
+			result.AppendNull()
+			continue
+		}
+		result.AppendString(string(cipherText))
```
Comment (XuHuaiyu) [inline expression/builtin_encryption_vec.go]: AppendString ==> AppendBytes

File: expression/builtin_encryption_vec.go (hunk 3 above)
Comment (js00070) [reply expression/builtin_encryption_vec.go]: It's very werid that the code will not pass the unit test if I replace the `GetString` with `GetBytes`. Maybe there are potential bugs in the code. I will try to debug it.

Comment (XuHuaiyu) [reply expression/builtin_encryption_vec.go]: ok, thanks

**Comment by js00070** [comment]
@XuHuaiyu 
The function `aesEncrypt` will change the memory of `str`. If we use `GetBytes`, the data layout in `chunk.Column` will be damanged. This is the result why the code can't pass the unit tests after replacing `GetString` with `GetBytes`
 
https://github.com/pingcap/tidb/blob/5a00578e2769ba0b8d1e9e2cb930de92ed6442b5/util/encrypt/aes.go#L240

File: expression/builtin_encryption_vec.go [hunk 5]
```diff
@@ -366,11 +366,62 @@ func (b *builtinCompressSig) vecEvalString(input *chunk.Chunk, result *chunk.Col
...
+		if strBuf.IsNull(i) || keyBuf.IsNull(i) {
+			result.AppendNull()
+			continue
+		}
+		if isWarning {
+			b.ctx.GetSessionVars().StmtCtx.AppendWarning(errWarnOptionIgnored.GenWithStackByArgs("IV"))
+		}
+		key := encrypt.DeriveKeyMySQL(keyBuf.GetBytes(i), b.keySize)
+
+		// we can't use GetBytes, because in AESEncryptWithECB padding is automatically
```
Comment (XuHuaiyu) [inline expression/builtin_encryption_vec.go]: It will be better to add `NOTE: xxx" comment for aesEncrypt, too.

**Pull request merged**